docs/source/hyperparameters.rst

```diff
@@ -152,25 +152,8 @@ In that case, choose only a few
     model = LitMNIST.load_from_checkpoint(PATH, loss_fx=torch.nn.SomeOtherLoss, generator_network=MyGenerator())
 
 
-3. Assign to `self.hparams`. Anything assigned to `self.hparams` will also be saved automatically
```

Comment:
but this still works right?

Reply:
I think this doc PR is meant to guide users towards best practices, but yeah it works (just problematic)

Reply:
@s-rog what's the problematic case where it won't work?

Reply:
The issue linked above (#4333) with saving/loading

Reply:
hmm, not sure. what do you guys recommend? warning or remove?

Reply:
Can we actually stop users from assigning via `self.hparams`? If not then perhaps a warning is better

Reply:
Do I correctly understand that code lines like `self.hparams["output_neurons"] = 3` should be rewritten to `self.save_hyperparameters({"output_neurons": 3})`?

How should cases where you append something to a list be treated? Before it was a simple line like `self.hparams["some_list"].append(some_value)`, should that become `self.save_hyperparameters(self.hparams["some_list"].append(some_value))`? I can image this could create performance issues?

Or an approach like this is better?:
```python
some_list = []
for x in foo_iterable:
    some_list.append(x)
self.save_hyperparameters(some_list)
```
This would risk forgetting to make this last `save_hyperparameters()` call however, creating a model .ckpt that can not be loaded back.

Sorry for the questions. In my code, dealing with Transfer Learning and I've been using a lot of `self.hparams` asssignments. I need to do things like:
```
model = FooModel()
model.load_pretrained_weights()
model.do_some_hparam_modifications()
```
to allow for both training from scratch, using pretrained weights for a new task, and load that new task after training, with the same code base.

Reply:
`self.hparams["some_list"].append(some_value)` should still work

Reply:
@s-rog Yes, but this is not automatically saved to `hparams.yaml` as described in this issue: https://github.com/PyTorchLightning/pytorch-lightning/issues/4316

Reply:
Also, what is the recommended way to set a new variable in `self.hparams` outside the `__init__`? `self.save_hyperparameters()` will throw the error `AssertionError: failed to inspect the self init` when used outside init, as can be seen in this colab: https://colab.research.google.com/drive/1Bc_R2XnR6_jOqmevXeaOJ3qxJ5ccqH6x?usp=sharing

Reply:
The answer to all your questions in your first comment above is no.
This is clearly not the way to use save_hyperparameters. If you change the params after you pass them into the model and then save them, what's the point? The point of save_hyperparameters is to save them exactly as they are passed into the model so that they can be loaded for reproducibility. 

> Also, what is the recommended way to set a new variable in self.hparams outside the __init__?

self.hparams is a dict, you can use it like a dict and change values.
If there is an issue with saving to hparams.yaml, let's discuss it on the github issue.

Reply:
@s-rog @rohitgr7 see latest docs build 
https://67894-178626720-gh.circle-artifacts.com/0/html/hyperparameters.html#lightningmodule-hyperparameters
I added a warning instead of removing it completely.

Reply:
> If you change the params after you pass them into the model and then save them, what's the point?

In this forum post I describe a transfer learning setup of which I cannot think of a solution how to satisfy all my use-cases without assigning directly to self.hparams:
https://forums.pytorchlightning.ai/t/how-to-save-hparams-when-not-provided-as-argument-apparently-assigning-to-hparams-is-not-recomended/339

I would love to know how I can support my use-cases if assigning to `self.hparams` directly would have been removed.